### PR TITLE
PS-1784 Add system metadata tags to files

### PR DESCRIPTION
**Pull request by pivnicek**
https://keboola.atlassian.net/browse/PS-1784

### CODE REVIEW

File: src/Writer/FileWriter.php [hunk 1]
```diff
@@ -5,6 +5,7 @@
 use Keboola\InputMapping\Reader;
 use Keboola\OutputMapping\Configuration\File\Manifest as FileManifest;
 use Keboola\OutputMapping\Exception\InvalidOutputException;
+use Keboola\OutputMapping\Exception\OutputOperationException;
```
Comment (odinuv) [inline src/Writer/FileWriter.php:8]: is it used?

File: src/Writer/Helper/TagsRewriter.php [hunk 2]
```diff
@@ -17,4 +17,14 @@ public static function rewriteTags(array $storageConfig, ClientWrapper $clientWr
 
         return $storageConfig;
     }
+
+    public static function addSystemTags(array $storageConfig, array $systemMetadata)
+    {
+        if (!empty($systemMetadata)) {
+            foreach ($systemMetadata as $systemKey => $systemValue) {
+                $storageConfig['tags'][] = $systemKey . ': ' . $systemValue;
```
Comment (odinuv) [inline src/Writer/Helper/TagsRewriter.php:25]: please name and enumerate them explicitly, this is gonna be huge chaos if half of the metadata is defined here https://github.com/keboola/output-mapping/blob/a53d90913686d8b0916855e3cef7e7ba430698de/src/Writer/TableWriter.php#L34 and the other half would be defined in docker-bundle
also component id is required https://github.com/keboola/output-mapping/blob/a53d90913686d8b0916855e3cef7e7ba430698de/src/Writer/TableWriter.php#L68 and the others are optional https://github.com/keboola/output-mapping/blob/a53d90913686d8b0916855e3cef7e7ba430698de/src/Writer/TableWriter.php#L557

File: src/Writer/FileWriter.php [hunk 3]
```diff
@@ -85,6 +87,7 @@ public function uploadFiles($source, $configuration, $storage)
                 );
             }
             try {
+                $storageConfig = TagsRewriter::addSystemTags($storageConfig, $systemMetadata);
```
Comment (odinuv) [inline src/Writer/FileWriter.php:90]: it doesn't really rewrite tags, does it?

File: tests/Writer/AbsWriterWorkspaceTest.php [hunk 4]
```diff
@@ -334,13 +341,22 @@ public function testWriteBasicFiles()
...
+        $file2expectedTags = array_merge($expectedTags, ['another-tag']);
+
         $this->assertNotNull($file1);
         $this->assertNotNull($file2);
         $this->assertNotNull($file3);
         $this->assertEquals(4, $file1['sizeBytes']);
-        $this->assertEquals(['output-mapping-test'], $file1['tags']);
-        $this->assertEquals(['output-mapping-test', 'another-tag'], $file2['tags']);
-        $this->assertEquals(['output-mapping-test'], $file3['tags']);
+        $this->assertEquals($expectedTags, $file1["tags"]);
```
Comment (odinuv) [inline tests/Writer/AbsWriterWorkspaceTest.php:357]: single quotes pls

File: tests/Writer/AbsWriterWorkspaceTest.php [hunk 5]
```diff
@@ -334,13 +341,22 @@ public function testWriteBasicFiles()
             }
         }
 
+        $expectedTags = [
+            'output-mapping-test',
+            'componentId: testComponent',
+            'configurationId: metadata-write-test',
+            'rowId: 12345',
```
Comment (odinuv) [inline tests/Writer/AbsWriterWorkspaceTest.php:348]: in reality, this has to be `configurationRowId` https://github.com/keboola/ex-google-bigquery-v2/blob/524a0f220cbf3fa1c1d6275859c1dbb86ced8899/src/OutputMappingStrategy/Files.php#L56 
which ofc makes no difference here, but `rowId: 12345` simply should not be produced

**Review by odinuv — CHANGES_REQUESTED** [review]

File: src/Writer/FileWriter.php (hunk 3 above)
Comment (pivnicek) [reply src/Writer/FileWriter.php:90]: no, but it manipulates tags, so I guess I can either change the class name or make a new class

File: src/Writer/Helper/TagsRewriter.php (hunk 2 above)
Comment (pivnicek) [reply src/Writer/Helper/TagsRewriter.php:25]: lol 
image

File: src/Writer/Helper/TagsRewriter.php [hunk 6]
```diff
@@ -17,4 +18,27 @@ public static function rewriteTags(array $storageConfig, ClientWrapper $clientWr
 
         return $storageConfig;
     }
+
+    public static function addSystemTags(array $storageConfig, array $systemMetadata)
+    {
+        if (!empty($systemMetadata)) {
```
Comment (odinuv) [inline src/Writer/Helper/TagsRewriter.php:24]: condition not needed (also always true in reality)

File: src/Writer/Helper/TagsRewriter.php [hunk 7]
```diff
@@ -17,4 +18,27 @@ public static function rewriteTags(array $storageConfig, ClientWrapper $clientWr
...
+            foreach ($systemMetadata as $systemKey => $systemValue) {
+                switch ($systemKey) {
```
Comment (odinuv) [inline src/Writer/Helper/TagsRewriter.php:26]: ok, so if you did it this way then, you can actually check it via in_array($systemkey, [FileWriter::SYSTEM_KEY_COMPONENT_ID, ...] and else log something like "key XY ignored"

**Comment by pivnicek** [comment]
right on, thanks man, I'll update accordingly

File: tests/Writer/StorageApiWriterTest.php [hunk 8]
```diff
@@ -21,6 +21,8 @@ class StorageApiWriterTest extends BaseWriterTest
 {
     use CreateBranchTrait;
 
+    const DEFAULT_SYSTEM_METADATA = ['componentId' => 'foo'];
```
Comment (odinuv) [inline tests/Writer/StorageApiWriterTest.php:24]: not used

**Comment by odinuv** [comment]
seems ok (when the tests pass)

File: tests/Writer/StorageApiWriterTest.php [hunk 9]
```diff
@@ -159,7 +177,14 @@ public function testWriteFilesOutputMappingDevMode()
 
         $writer = new FileWriter($this->getStagingFactory());
 
-        $writer->uploadFiles('/upload', ["mapping" => $configs], StrategyFactory::LOCAL);
+        $systemMetadata = [
+            'componentId' => 'testComponent',
+            'configurationId' => 'metadata-write-test',
+            'configurationRowId' => '7998-configurationRowId: 12345',
```
Comment (odinuv) [inline tests/Writer/StorageApiWriterTest.php:183]: a bit strange value for configuration row id

Comment (pivnicek) [reply tests/Writer/StorageApiWriterTest.php:183]: yeah, that's not right at all :-) thanks

Comment (odinuv) [reply tests/Writer/StorageApiWriterTest.php:183]: it's an opaque string, so in a way it doesn't matter. Looks weird though :)

**Review by odinuv — APPROVED** [review]

**Pull request merged**